HLTrigger/Timer/interface/FastTimerService.h

```diff
@@ -86,690 +82,488 @@ class FastTimerService {
   FastTimerService(const edm::ParameterSet &, edm::ActivityRegistry & );
   ~FastTimerService();
 
-  // reserve plots for timing vs. luminosity - these should only be called before any begin run (actually, before any preStreamBeginRun)
-  unsigned int reserveLuminosityPlots(std::string const & name, std::string const & title, std::string const & label, double range, double resolution);
-  unsigned int reserveLuminosityPlots(std::string && name, std::string && title, std::string && label, double range, double resolution);
-
-  // set the event luminosity
-  void setLuminosity(unsigned int stream_id, unsigned int luminosity_id, double value);
-
-  // query the current module/path/event
-  // Note: these functions incur in a "per-call timer overhead" (see above), currently of the order of 340ns
-  double currentModuleTime(edm::StreamID) const;            // return the time spent since the last preModuleEvent() event
-  double currentPathTime(edm::StreamID) const;              // return the time spent since the last prePathEvent() event
-  double currentEventTime(edm::StreamID) const;             // return the time spent since the last preEvent() event
+private:
+  double queryModuleTime_(edm::StreamID, unsigned int id) const;
 
-  // query the time spent in a module/path (available after it has run)
-  double queryModuleTime(edm::StreamID, const edm::ModuleDescription &) const;
-  double queryModuleTime(edm::StreamID, unsigned int id) const;
-  double queryModuleTimeByLabel(edm::StreamID, const std::string &) const;
-  double queryModuleTimeByType(edm::StreamID, const std::string &) const;
-  /* FIXME re-implement taking into account subprocesses
-  double queryPathActiveTime(edm::StreamID, const std::string &) const;
-  double queryPathExclusiveTime(edm::StreamID, const std::string &) const;
-  double queryPathTotalTime(edm::StreamID, const std::string &) const;
-  */
-
-  // query the time spent in the current event's
-  //  - source        (available during event processing)
-  //  - all paths     (available during endpaths)
-  //  - all endpaths  (available after all endpaths have run)
-  //  - processing    (available after the event has been processed)
+public:
+  // query the time spent in a module/path/process (available after it has run)
```

Comment:
I'm concerned that the `query` calls may be causing undefined behavior since you have no explicit synchronization going on between the threads that are writing the data and those calling `query`. TBB tasks internally do a memory barrier (which is a synchronization) before the task starts and again after the task finishes. This guarantees that if a task depends on the results of another task which finished before it that the memory is all synchronized. As far as I can tell, your `query` calls have no such guarantees.